mutiny-core/src/federation.rs

```diff
@@ -0,0 +1,612 @@
+    get_config_from_db,
+    oplog::{OperationLogEntry, UpdateStreamOrOutcome},
+    secret::get_default_client_secret,
+    ClientArc, FederationInfo,
+};
+use fedimint_core::{
+    api::InviteCode,
+    config::FederationId,
+    core::OperationId,
+    task::{MaybeSend, MaybeSync},
+    Amount,
+};
+use fedimint_ln_client::{
+    InternalPayState, LightningClientInit, LightningClientModule, LightningOperationMeta,
+    LightningOperationMetaVariant, LnPayState, LnReceiveState,
+};
+use fedimint_mint_client::MintClientInit;
+use fedimint_wallet_client::WalletClientInit;
+use futures::future::{self};
+use futures_util::{pin_mut, stream, StreamExt};
+use lightning::log_info;
+use lightning::{log_warn, util::logger::Logger};
+use lightning_invoice::Bolt11Invoice;
+use serde::{de::DeserializeOwned, Deserialize, Serialize};
+use std::fmt::Debug;
+use std::{
+    collections::HashMap,
+    sync::{atomic::AtomicBool, Arc, RwLock},
+};
+
+const FEDIMINT_CLIENT_NONCE: &[u8] = b"Fedimint Client Salt";
+
+impl From<LnReceiveState> for HTLCStatus {
+    fn from(state: LnReceiveState) -> Self {
+        match state {
+            LnReceiveState::Created => HTLCStatus::Pending,
+            LnReceiveState::Claimed => HTLCStatus::Succeeded,
+            LnReceiveState::WaitingForPayment { .. } => HTLCStatus::Pending,
+            LnReceiveState::Canceled { .. } => HTLCStatus::Failed,
+            LnReceiveState::Funded => HTLCStatus::InFlight,
+            LnReceiveState::AwaitingFunds => HTLCStatus::InFlight,
+        }
+    }
+}
+
+impl From<InternalPayState> for HTLCStatus {
+    fn from(state: InternalPayState) -> Self {
+        match state {
+            InternalPayState::Funding => HTLCStatus::InFlight,
+            InternalPayState::Preimage(_) => HTLCStatus::Succeeded,
+            InternalPayState::RefundSuccess { .. } => HTLCStatus::Failed,
+            InternalPayState::RefundError { .. } => HTLCStatus::Failed,
+            InternalPayState::FundingFailed { .. } => HTLCStatus::Failed,
+            InternalPayState::UnexpectedError(_) => HTLCStatus::Failed,
+        }
+    }
+}
+
+impl From<LnPayState> for HTLCStatus {
+    fn from(state: LnPayState) -> Self {
+        match state {
+            LnPayState::Created => HTLCStatus::Pending,
+            LnPayState::Canceled => HTLCStatus::Failed,
+            LnPayState::Funded => HTLCStatus::InFlight,
+            LnPayState::WaitingForRefund { .. } => HTLCStatus::InFlight,
+            LnPayState::AwaitingChange => HTLCStatus::InFlight,
+            LnPayState::Success { .. } => HTLCStatus::Succeeded,
+            LnPayState::Refunded { .. } => HTLCStatus::Failed,
+            LnPayState::UnexpectedError { .. } => HTLCStatus::Failed,
+        }
+    }
+}
+
+// This is the FederationStorage object saved to the DB
+#[derive(Debug, Serialize, Deserialize, Clone, Default)]
+pub struct FederationStorage {
+    pub federations: HashMap<String, FederationIndex>,
+    #[serde(default)]
```

Comment:
shouldn't need this `#[serde(default)]`, only needed that for the node storage because we had ones without a version.

Might be better to just use `VersionedValue` instead so we don't need to add another case for serialization

Reply:
Idk about `VersionedValue`. We're only currently doing it for the channel manager, and I think that's mostly so that we can compare it's internal value with the value stored, right? Why don't we use it for channel monitor, nodes, locks, etc? 

Also, I would rather be passing around `FederationStorage` to multiple places instead of passing around and parsing a generic `Value`. We've got to parse it somewhere, and should be using real values. Plus we never know what we're going to add to this later.

Reply:
Not handling for now, will worry about this as we tighten up more storage stuff.